### PR TITLE
upgrades,jobs: stub the auto config runner

### PR DESCRIPTION
**Pull request by knz**
Epic: CRDB-23559
Informs #98431.

This change is part of a larger project to perform auto-configuration in newly created clusters. One of the components will be the "auto config runner", a job which receives configuration to apply from the environment, creates other jobs to apply this configuration, and stores the execution results.

This commit stubs the job, which means allocating a cluster version for it and instantiating it via a cluster migration. The job is otherwise idle at this point.

Release note: None

### CODE REVIEW

File: pkg/upgrade/upgrades/create_auto_config_runner_job.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+			nil,
+			sessiondata.InternalExecutorOverride{User: username.RootUserName()},
+			"SELECT * FROM system.jobs WHERE id = $1",
+			jobs.AutoConfigRunnerJobID,
+		)
+		if err != nil {
+			return err
+		}
+
+		if row == nil {
```
Comment (ajwerner) [inline pkg/upgrade/upgrades/create_auto_config_runner_job.go:45]: uber-nit: invert the conditional to early return and lose some indentation.

File: pkg/upgrade/upgrades/create_auto_config_runner_job.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+) error {
+	if d.TestingKnobs != nil && d.TestingKnobs.SkipAutoConfigRunnerJobBootstrap {
+		return nil
+	}
+	return d.DB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		row, err := d.DB.Executor().QueryRowEx(
+			ctx,
+			"check for existing auto config runner job",
+			nil,
+			sessiondata.InternalExecutorOverride{User: username.RootUserName()},
```
Comment (ajwerner) [inline pkg/upgrade/upgrades/create_auto_config_runner_job.go:37]: Two things here:
1) Why root and not node? (we are so unprincipled about it)
2) Want to just use the var in that package which is this value?

File: pkg/upgrade/upgrades/create_auto_config_runner_job.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+				NonCancelable: true,
+			}
+			// Use CreateJob instead of CreateAdoptableJob to ensure this node
+			// has a claim on the job and can start it immediately below.
+			if _, err := d.JobRegistry.CreateJobWithTxn(ctx, jr, jobs.AutoConfigRunnerJobID, txn); err != nil {
+				return err
+			}
+			// Start the job immediately. This speeds up the application
+			// of initial configuration tasks.
+			d.JobRegistry.NotifyToResume(ctx, jobs.AutoConfigRunnerJobID)
```
Comment (ajwerner) [inline pkg/upgrade/upgrades/create_auto_config_runner_job.go:62]: If you do this here, when the job doesn't yet exist (transaction hasn't committed), there's a chance you'll have to wait an entire adoption cycle. 

You should call this after the transaction commits.

File: pkg/server/autoconfig/auto_config.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package autoconfig
```
Comment (ajwerner) [inline pkg/server/autoconfig/auto_config.go:11]: depending on how the package structure comes together, you may want to move this into a subpackage called autoconfigjob or something. The jobs end up taking on a lot of dependencies. If there are other APIs you're going to want lower dependency packages to use, you'll be sad having the job in the package at the root of the subsystem.

**Review by ajwerner — APPROVED** [review]
Just minor stuff from me. Feel free to merge.

File: pkg/upgrade/upgrades/create_auto_config_runner_job.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	if d.TestingKnobs != nil && d.TestingKnobs.SkipAutoConfigRunnerJobBootstrap {
+		return nil
+	}
+	return d.DB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		row, err := d.DB.Executor().QueryRowEx(
+			ctx,
+			"check for existing auto config runner job",
+			nil,
+			sessiondata.InternalExecutorOverride{User: username.RootUserName()},
+			"SELECT * FROM system.jobs WHERE id = $1",
```
Comment (stevendanna) [inline pkg/upgrade/upgrades/create_auto_config_runner_job.go:38]: [nit] No need to return everything here.

**Review by stevendanna — APPROVED** [review]
One minor nit on top of Andrew's comments. Otherwise this looks reasonable.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98440)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @dt, and @stevendanna)

___
*[`pkg/server/autoconfig/auto_config.go` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98440#-NQRyMVXxAHGLHZKpg_--r3-11:-NQ_8Z8E4tQvGeo976IM:b-vc7pq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/039ecf129920d3762e0fdf3655a1ea542511bba2/pkg/server/autoconfig/auto_config.go#L11)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

depending on how the package structure comes together, you may want to move this into a subpackage called autoconfigjob or something. The jobs end up taking on a lot of dependencies. If there are other APIs you're going to want lower dependency packages to use, you'll be sad having the job in the package at the root of the subsystem. 
</blockquote></details>

I agree with this. I will do this in the next PR.
___
*[`pkg/upgrade/upgrades/create_auto_config_runner_job.go` line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98440#-NQRyMVXxAHGLHZKpg_0-r3-37:-NQ_8cEyD2CWcvXSkXSB:bjdemy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/039ecf129920d3762e0fdf3655a1ea542511bba2/pkg/upgrade/upgrades/create_auto_config_runner_job.go#L37)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Two things here:
1) Why root and not node? (we are so unprincipled about it)
2) Want to just use the var in that package which is this value?
</blockquote></details>

As to question (1) i simply copy-pasted this file from another similarly named file in the same directory. This is what the other file was doing. I have a small preference for 'root' for these things because 'node' does not have an entry in system.users and certain SHOW statements care about it.

Point (2) - done.
___
*[`pkg/upgrade/upgrades/create_auto_config_runner_job.go` line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98440#-NQRyMVXxAHGLHZKpg_0-r3-38:-NQ_9-kJ-xNdoJB07lBG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/039ecf129920d3762e0fdf3655a1ea542511bba2/pkg/upgrade/upgrades/create_auto_config_runner_job.go#L38)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] No need to return everything here.
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/create_auto_config_runner_job.go` line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98440#-NQRyMVXxAHGLHZKpg_0-r3-45:-NQ_98cn3UevzjjjQTHg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/039ecf129920d3762e0fdf3655a1ea542511bba2/pkg/upgrade/upgrades/create_auto_config_runner_job.go#L45)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

uber-nit: invert the conditional to early return and lose some indentation. 
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/create_auto_config_runner_job.go` line 62 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98440#-NQRyMVXxAHGLHZKpg_0-r3-62:-NQ_9HrhDIx1a6TOfEG5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/039ecf129920d3762e0fdf3655a1ea542511bba2/pkg/upgrade/upgrades/create_auto_config_runner_job.go#L62)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

If you do this here, when the job doesn't yet exist (transaction hasn't committed), there's a chance you'll have to wait an entire adoption cycle. 

You should call this after the transaction commits.
</blockquote></details>

Done.

**Comment by knz** [comment]
bors r=stevendanna,ajwerner

**Comment by knz** [comment]
bors r=stevendanna,ajwerner

**Pull request merged**